### PR TITLE
updates to WalkerControl to have useful walker_ids

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

Walker ID's in batched version of the drivers should now supply trackable walker ID's

supercedes #5010

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation or build script changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
OSX laptop
x86 build

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes/No. This PR is up to date with current the current state of 'develop'
- Yes/No. Code added or changed in the PR has been clang-formatted
- Yes/No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes/No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by prckent** [comment]
Test this please

**Comment by PDoakORNL** [comment]
This is ready for review

File: src/QMCDrivers/DMC/WalkerControl.cpp [hunk 1]
```diff
@@ -236,10 +225,16 @@ void WalkerControl::branch(int iter, MCPopulation& pop, bool do_not_branch)
     throw std::runtime_error("Potential bug! Population num_global_walkers mismatched!");
 #endif
 
+  // If we are in warmup it think it should be a fatal error is walker->Multiplicity != 1.0
```
Comment (ye-luo) [inline src/QMCDrivers/DMC/WalkerControl.cpp:228]: DMC warmup follow normal branching. Only the first iteration flags do_not_branch.

File: src/QMCDrivers/DMC/WalkerControl.cpp [hunk 2]
```diff
@@ -27,6 +27,8 @@
 #include "type_traits/template_types.hpp"
 #include "QMCWaveFunctions/TrialWaveFunction.h"
 
+#include "Utilities/for_testing/NativeInitializerPrint.hpp"
```
Comment (ye-luo) [inline src/QMCDrivers/DMC/WalkerControl.cpp:30]: Either move it out of the test space as a generic feature or avoid using it.

File: src/QMCDrivers/DMC/WalkerControl.h [hunk 3]
```diff
@@ -152,7 +165,12 @@ class WalkerControl : public MPIObjectBase
   IndexType max_copy_;
   ///trial energy energy
   FullPrecRealType trial_energy_;
-  ///number of walkers on each MPI rank after branching before load balancing
+  /** number of walkers on each MPI wrt to the branching algorithm before load balancing
+   *  not the same as WalkerControl::branch.
```
Comment (ye-luo) [inline src/QMCDrivers/DMC/WalkerControl.h:169]: Could you clarify what this line means?

File: src/QMCDrivers/tests/test_WalkerControl.cpp [hunk 4]
```diff
@@ -48,24 +47,24 @@ UnifiedDriverWalkerControlMPITest::UnifiedDriverWalkerControlMPITest() : wc_(dpo
...
+//     {
+//       we.walker.registerData();
+//       we.twf.registerData(we.pset, we.walker.DataSet);
+//       we.walker.DataSet.allocate();
+//     }
+//     we.twf.copyFromBuffer(we.pset, we.walker.DataSet);
+//     we.twf.evaluateLog(we.pset);
+//     we.twf.updateBuffer(we.pset, we.walker.DataSet);
+//   }
+// }
```
Comment (ye-luo) [inline src/QMCDrivers/tests/test_WalkerControl.cpp:67]: Remove?

File: src/QMCDrivers/tests/test_WalkerControl.cpp [hunk 5]
```diff
@@ -79,93 +78,266 @@ void UnifiedDriverWalkerControlMPITest::reportWalkersPerRank(Communicate* c, MCP
...
-    for (int i = 0; i < rank_walker_count.size(); ++i)
-    {
-      std::cout << " " << i << "  " << rank_walker_count[i] << '\n';
-    }
+    // std::cout << "Rank   Count\n"
+    //    << "===========\n";
+    // for (int i = 0; i < rank_walker_count.size(); ++i)
+    // {
+    //   std::cout << std::setw(4) << i << "   " << rank_walker_count[i] << '\n';
+    // }
```
Comment (ye-luo) [inline src/QMCDrivers/tests/test_WalkerControl.cpp:86]: you can print them to `app_log()` which are hidden by default.

File: src/QMCDrivers/DMC/WalkerControl.cpp [hunk 6]
```diff
@@ -134,23 +136,27 @@ void WalkerControl::writeDMCdat(int iter, const std::vector<FullPrecRealType>& c
   }
 }
 
+/** unified: perform branch and swap (balance) walkers as required
+   *  **This has many side effects**
+   *  ## For:
+   *  ### dynamic population
+   *  1. compute multiplicity. If iter 0 and all of warmup -> multiplicity = 1
```
[the remark this answers is not in the document]
Comment (PDoakORNL) [reply src/QMCDrivers/DMC/WalkerControl.cpp:143]:
```
      if (do_not_branch)
        for (auto& walker : walkers)
          walker->Multiplicity = 1.0;
```
Thats my understanding of this code around line 181.

File: src/QMCDrivers/DMC/WalkerControl.cpp (hunk 1 above)
Comment (ye-luo) [reply src/QMCDrivers/DMC/WalkerControl.cpp:228]: Remove this comment?

File: src/QMCDrivers/DMC/WalkerControl.cpp (hunk 2 above)
Comment (ye-luo) [reply src/QMCDrivers/DMC/WalkerControl.cpp:30]: ?

**Review by ye-luo — COMMENTED** [review]
Please address the two remaining conversation. On a side note, I ran a DMC simulation and everything is fine.

File: src/QMCDrivers/DMC/WalkerControl.cpp (hunk 2 above)
Comment (PDoakORNL) [reply src/QMCDrivers/DMC/WalkerControl.cpp:30]: removing, sorry thought I'd pushed this.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**